sqlite/ext/comdb2/threadpools.c

```diff
@@ -0,0 +1,288 @@
+    COLUMN_QUEUE,
+    COLUMN_LONG_WAIT_MS,
+    COLUMN_LINGER_SECS,
+    COLUMN_STACK_SIZE,
+    COLUMN_MAX_QUEUE_OVERRIDE,
+    COLUMN_MAX_QUEUE_AGE_MS,
+    COLUMN_EXIT_ON_CREATE_FAIL,
+    COLUMN_DUMP_ON_FULL,
+    /*COLUMN_HISTOGRAM,*/
+};
+
+static int systblThreadPoolsConnect(sqlite3 *db, void *pAux, int argc,
+                                    const char *const *argv,
+                                    sqlite3_vtab **ppVtab, char **pErr)
+{
+    int rc;
+
+    rc = sqlite3_declare_vtab(
+        db, "CREATE TABLE comdb2_pools(\"name\", \"status\", \"num_thd\", "
+            "\"free_thd\", \"peak_thd\", \"num_creates\", \"num_exits\", "
+            "\"num_passed\", \"num_enqueued\", \"num_dequeued\", "
+            "\"num_timeout\", \"num_failed_dispatches\", \"min_thds\", "
+            "\"max_thds\", \"peak_queue\", \"max_queue\", \"queue\", "
+            "\"long_wait_ms\", \"linger_secs\", \"stack_size\", "
+            "\"max_queue_override\", \"max_queue_age_ms\", "
+            "\"exit_on_create_fail\", \"dump_on_full\")");
+
+    if (rc == SQLITE_OK) {
+        if ((*ppVtab = sqlite3_malloc(sizeof(sqlite3_vtab))) == 0) {
+            return SQLITE_NOMEM;
+        }
+        memset(*ppVtab, 0, sizeof(*ppVtab));
+    }
+
+    return 0;
+}
+
+static int systblThreadPoolsBestIndex(sqlite3_vtab *tab,
+                                      sqlite3_index_info *pIdxInfo)
+{
+    return SQLITE_OK;
+}
+
+static int systblThreadPoolsDisconnect(sqlite3_vtab *pVtab)
+{
+    sqlite3_free(pVtab);
+    return SQLITE_OK;
+}
+
+static int systblThreadPoolsOpen(sqlite3_vtab *p,
+                                 sqlite3_vtab_cursor **ppCursor)
+{
+    systbl_threadpools_cursor *cur =
+        sqlite3_malloc(sizeof(systbl_threadpools_cursor));
+    if (cur == 0) {
+        return SQLITE_NOMEM;
+    }
+    memset(cur, 0, sizeof(*cur));
+    *ppCursor = &cur->base;
+
+    /* Unlocked in systblThreadPoolsClose() */
+    pthread_mutex_lock(&pool_list_lk);
+    cur->pool = LISTC_TOP(&threadpools);
+
+    return SQLITE_OK;
+}
+
+static int systblThreadPoolsClose(sqlite3_vtab_cursor *cur)
+{
+    pthread_mutex_unlock(&pool_list_lk);
+    sqlite3_free(cur);
+    return SQLITE_OK;
+}
+
+static int systblThreadPoolsFilter(sqlite3_vtab_cursor *pVtabCursor, int idxNum,
+                                   const char *idxStr, int argc,
+                                   sqlite3_value **argv)
+{
+    systbl_threadpools_cursor *pCur = (systbl_threadpools_cursor *)pVtabCursor;
+    pCur->rowid = 0;
+    return SQLITE_OK;
+}
+
+static int systblThreadPoolsNext(sqlite3_vtab_cursor *cur)
+{
+    systbl_threadpools_cursor *pCur = (systbl_threadpools_cursor *)cur;
+    pCur->pool = thdpool_next_pool(pCur->pool);
+    pCur->rowid++;
+    return SQLITE_OK;
+}
+
+static int systblThreadPoolsEof(sqlite3_vtab_cursor *cur)
+{
+    systbl_threadpools_cursor *pCur = (systbl_threadpools_cursor *)cur;
+    return (pCur->rowid >= listc_size(&threadpools)) ? 1 : 0;
```

Comment:
You only need to `return (pCur->pool == NULL)` here as `thdpool_next_pool()` returns `NULL` if no next node.